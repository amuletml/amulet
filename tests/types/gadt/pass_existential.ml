external val print : string -> unit = "print"
external val io_write : string -> unit = "io.write"

type natural = Z | S of natural

type nat 'n =
  | Zero : nat Z
  | Succ : nat 'k -> nat (S 'k)

type some_nat =
  | SomeNat : nat 'n -> some_nat

let with_natural i (k : forall 'n. nat 'n -> 'b) =
  let go n =
    if n == 0 then
      SomeNat Zero
    else
      match go (n - 1) with
      | SomeNat x -> SomeNat (Succ x)
  in match go i with
  | SomeNat n -> k n

let main = with_natural 10 @@ fun x ->
  with_natural 20 @@ fun y ->
    let ppr_nat (n : nat 'n) : unit =
      match n with
      | Zero -> print "Zero"
      | Succ k ->
        io_write "Succ ";
        ppr_nat k
    ppr_nat x
    ppr_nat y
