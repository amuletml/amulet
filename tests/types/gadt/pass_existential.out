print : string -> unit
io_write : string -> unit
- : int -> int -> int
== : int -> int -> bool
natural : type
Z : natural
S : natural -> natural
nat : natural -> type
Zero : Infer{'cx : type}. Spec{'n : 'cx}. ('n ~ Z) âŠƒ nat 'n
Succ : Infer{'du : type}. Spec{'n : 'du}. Spec{'k : natural}. ('n ~ S 'k) âŠƒ nat 'k -> nat 'n
some_nat : type
SomeNat : Spec{'n : natural}. nat 'n -> some_nat
with_natural : Infer{'a : type}. int -> (Spec{'n : natural}. nat 'n -> 'a) -> 'a
main : unit
